tests/test_ag_ui.py

```diff
@@ -0,0 +1,941 @@
+                '{"type":"RUN_FINISHED","threadId":"thread_00000000-0000-0000-0000-000000000001","runId":"run_00000000-0000-0000-0000-000000000002"}',
+            ],
+        ),
+        AdapterRunTest(
+            id='thinking',
+            runs=[
+                Run(
+                    nodes=[
+                        TestNode(
+                            parts=[
+                                TestThinkingPart(content='Thinking'),
+                                TestThinkingPart(content='Thinking about the weather'),
+                                TestTextPart('Thought about the weather'),
+                            ],
+                        ),
+                    ],
+                    messages=[  # pyright: ignore[reportArgumentType]
+                        UserMessage(
+                            id='msg_1',
+                            role=Role.USER.value,
+                            content='Hello',
+                        ),
+                    ],
+                ),
+            ],
+            expected_events=[
+                '{"type":"RUN_STARTED","threadId":"thread_00000000-0000-0000-0000-000000000001","runId":"run_00000000-0000-0000-0000-000000000002"}',
+                '{"type":"THINKING_TEXT_MESSAGE_START"}',
+                '{"type":"THINKING_TEXT_MESSAGE_CONTENT","delta":"\\"Thinking\\""}',
+                '{"type":"THINKING_TEXT_MESSAGE_END"}',
+                '{"type":"THINKING_TEXT_MESSAGE_START"}',
+                '{"type":"THINKING_TEXT_MESSAGE_CONTENT","delta":"\\"Thinking "}',
+                '{"type":"THINKING_TEXT_MESSAGE_CONTENT","delta":"about "}',
+                '{"type":"THINKING_TEXT_MESSAGE_CONTENT","delta":"the "}',
+                '{"type":"THINKING_TEXT_MESSAGE_CONTENT","delta":"weather\\""}',
+                '{"type":"THINKING_TEXT_MESSAGE_END"}',
+                '{"type":"TEXT_MESSAGE_START","messageId":"00000000-0000-0000-0000-000000000003","role":"assistant"}',
+                '{"type":"TEXT_MESSAGE_CONTENT","messageId":"00000000-0000-0000-0000-000000000003","delta":"Thought "}',
+                '{"type":"TEXT_MESSAGE_CONTENT","messageId":"00000000-0000-0000-0000-000000000003","delta":"about "}',
+                '{"type":"TEXT_MESSAGE_CONTENT","messageId":"00000000-0000-0000-0000-000000000003","delta":"the "}',
+                '{"type":"TEXT_MESSAGE_CONTENT","messageId":"00000000-0000-0000-0000-000000000003","delta":"weather"}',
+                '{"type":"TEXT_MESSAGE_END","messageId":"00000000-0000-0000-0000-000000000003"}',
+                '{"type":"RUN_FINISHED","threadId":"thread_00000000-0000-0000-0000-000000000001","runId":"run_00000000-0000-0000-0000-000000000002"}',
+            ],
+        ),
+        AdapterRunTest(
+            id='tool_local_then_ag_ui',
+            call_tools=['current_time', 'get_weather'],
+            runs=[
+                Run(
+                    nodes=[
```

Comment:
I think the `custom_response_nodes` stuff is a lot to add to `TestModel` for these 2 uses of `Run(nodes=...)` that I think we could replace with using `FunctionModel` and a manually built `ModelResponse`